File: benchmarks/perf/dispatch.c
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

typedef int64_t (*ApplyFn)(int64_t amount, int64_t value);

typedef struct {
    int64_t amount;
    ApplyFn apply;
} Operation;

static inline int64_t checkedAdd(int64_t left, int64_t right) {
    int64_t result;
    if (__builtin_add_overflow(left, right, &result)) {
        abort();
    }
    return result;
}

static int64_t applyOffset(int64_t amount, int64_t value) {
    return checkedAdd(amount, value);
}

static int64_t run(Operation operation) {
    int64_t total = 0;
    for (int64_t value = 0; value < 100000000; ++value) {
        total = checkedAdd(total, operation.apply(operation.amount, value));
    }
    return total;
}

int main(void) {
    Operation operation = {.amount = 1, .apply = applyOffset};
    printf("%" PRId64 "\n", run(operation));
    return 0;
}

File: benchmarks/perf/dispatch.rs
trait Operation {
    fn apply(&self, value: i64) -> i64;
}

struct Offset {
    amount: i64,
}

impl Operation for Offset {
    fn apply(&self, value: i64) -> i64 {
        self.amount.checked_add(value).expect("benchmark overflow")
    }
}

fn run(operation: &dyn Operation) -> i64 {
    let mut total = 0_i64;
    for value in 0_i64..100_000_000 {
        total = total
            .checked_add(operation.apply(value))
            .expect("benchmark overflow");
    }
    total
}

fn main() {
    let operation = Offset { amount: 1 };
    println!("{}", run(&operation));
}
